base/distributed/cluster.jl

```diff
@@ -210,13 +210,27 @@ end
 # The master process uses this to connect to the worker and subsequently
 # setup a all-to-all network.
 function read_worker_host_port(io::IO)
-    while true
-        conninfo = readline(io)
+    t0 = time()
+    timeout = worker_timeout()
+
+    ntries = 1000
+    while ntries > 0
+        readtask = @schedule readline(io)
+        yield()
+        while !istaskdone(readtask) && ((time() - t0) < timeout)
+            sleep(0.05)
+        end
+        !istaskdone(readtask) && break
+
+        conninfo = wait(readtask)
+        ntries -= 1
         bind_addr, port = parse_connection_info(conninfo)
         if bind_addr != ""
             return bind_addr, port
         end
     end
+    close(io)
+    error("Error trying to read worker host:port from worker")
```

Comment:
The user gets to see the cause since the ssh command prints an error message of the form `ssh: Could not resolve hostname a: nodename nor servname provided, or not known`. However that is not known in this function here - it could be any other ssh error, an error running the julia command or a timeout.

This PR betters the current behavior. Identifying the root cause which requires a bit more restructuring can be done in a separate PR.